src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadNotificationService.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.HotReload;
+
+/// <summary>
+/// MEF Export that can be used to be notified when a project enters and exits hot reload 
+/// </summary>
+[Export(typeof(IProjectHotReloadNotificationService))]
+internal class ProjectHotReloadNotificationService : IProjectHotReloadNotificationService
+{
+    [ImportingConstructor]
+    public ProjectHotReloadNotificationService(UnconfiguredProject _)
+    {
+    }
+
+    public event AsyncEventHandler<bool>? HotReloadStateChangedAsync;
+
+    public bool IsProjectInHotReload { get; private set; }
+
+    public async Task SetHotReloadStateAsync(bool isInHotReload)
+    {
+        IsProjectInHotReload = isInHotReload;
+
+        var localEvent = HotReloadStateChangedAsync;
+        if (localEvent is not null)
+        {
+            try
+            {
+                await localEvent.InvokeAsync(this, isInHotReload);
+            }
+            catch (Exception ex)
+            {
+                System.Diagnostics.Debug.Fail($"HotReloadStartChanged handler threw an exception {ex}");
```

Comment:
Ideally the type would import the `IProjectFaultHandlerService` and report the failure there rather than calling `Debug.Fail`. I'm fine with that being cleaned up in a future PR though.

Reply:
Looking at the documentation for the that interface, I wouldn't want this to be visible to the user.  I'll leave this change to a future PR cleanup.